# federated learning with flower

same problem as the baseline notebook, but each device trains locally now and only
sends model weights to a central server, not raw data. this is actually the point
of the project - checking what you give up (if anything) by going federated instead
of centralized

In [1]:
!pip install -q flwr

import flwr as fl
print("flwr version:", fl.__version__)
import torch
import torch.nn as nn
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import roc_auc_score
from collections import OrderedDict

device_torch = "cpu"
torch.manual_seed(0)
np.random.seed(0)

from google.colab import drive
drive.mount('/content/drive')

df = pd.read_csv("/content/drive/MyDrive/nbaiot-project/processed/combined_labeled.csv")
feature_cols = [c for c in df.columns if c not in ["label", "attack_type", "device"]]
DEVICES = df["device"].unique().tolist()
DEVICES

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 25.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 268.7/268.7 kB 8.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.5/4.5 MB 36.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 129.3/129.3 kB 5.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 66.5/66.5 kB 2.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.3/57.3 kB 1.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 323.4/323.4 kB 7.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.3/2.3 MB 34.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 73.1/73.1 kB 1.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 310.5/310.5 kB 12.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 73.6/73.6 kB 1.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 251.6/251.6 kB 6.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 47.4/47.4 kB

['Danmini_Doorbell',
 'Ecobee_Thermostat',
 'Philips_B120N10_Baby_Monitor',
 'SimpleHome_XCS7_1002_WHT_Security_Camera']

same autoencoder as the baseline notebook, keeping it identical so the comparison
at the end is actually fair

In [2]:
class Autoencoder(nn.Module):
    def __init__(self, input_dim, latent_dim=16):
        super().__init__()
        self.encoder = nn.Sequential(
            nn.Linear(input_dim, 64), nn.ReLU(),
            nn.Linear(64, latent_dim), nn.ReLU(),
        )
        self.decoder = nn.Sequential(
            nn.Linear(latent_dim, 64), nn.ReLU(),
            nn.Linear(64, input_dim),
        )
    def forward(self, x):
        return self.decoder(self.encoder(x))

INPUT_DIM = len(feature_cols)
# shared scaler from TRAIN rows only (first 70% of each device's benign data, file order),
# built from per-client count / sum / sum of squares so no raw rows are pooled
class FedScaler:
    def __init__(self, mean, scale):
        self.mean, self.scale = mean, scale
    def transform(self, X):
        return (X - self.mean) / self.scale

def build_fed_scaler():
    n, s, ss = 0, 0.0, 0.0
    for dev in DEVICES:
        d = df[(df["device"] == dev) & (df["label"] == 0)][feature_cols].values.astype(np.float64)
        d = d[:int(len(d) * 0.7)]
        n += len(d); s = s + d.sum(0); ss = ss + (d ** 2).sum(0)
    mean = s / n
    scale = np.sqrt(np.maximum(ss / n - mean ** 2, 0))
    scale[scale < 1e-8] = 1.0
    return FedScaler(mean, scale)

global_scaler = build_fed_scaler()

In [3]:
import os, gc

CLIENT_DIR = "/content/client_data"
os.makedirs(CLIENT_DIR, exist_ok=True)

def prepare_client_files():
    """Write each device's scaled train/test arrays to its own file, so a simulated
    client loads only its own data instead of a copy of the full dataframe."""
    for dev in DEVICES:
        d = df[df["device"] == dev]
        benign = d[d["label"] == 0][feature_cols].values
        attack = d[d["label"] == 1][feature_cols].values
        n_train = int(len(benign) * 0.7)          # first 70% in file order
        test_benign = benign[n_train:]
        X_test = np.vstack([test_benign, attack]) if len(attack) else test_benign
        y_test = np.hstack([np.zeros(len(test_benign)), np.ones(len(attack))])
        np.savez(
            f"{CLIENT_DIR}/{dev}.npz",
            train=global_scaler.transform(benign[:n_train]).astype(np.float32),
            X_test=global_scaler.transform(X_test).astype(np.float32),
            y_test=y_test,
        )
        print(dev, "train", n_train, "test", len(X_test))

prepare_client_files()
del df          # clients no longer reference the full dataframe
gc.collect()

def get_client_data(device_name):
    z = np.load(f"{CLIENT_DIR}/{device_name}.npz")
    return z["train"], z["X_test"], z["y_test"]

Danmini_Doorbell train 34683 test 313531
Ecobee_Thermostat train 9179 test 243735
Philips_B120N10_Baby_Monitor train 122667 test 333818
SimpleHome_XCS7_1002_WHT_Security_Camera train 32609 test 254248


## flower client

each simulated client = one iot device. trains on its own local benign traffic,
sends weights back, never sends raw data anywhere

In [4]:
# standard flwr NumPyClient setup, fit() trains locally, evaluate() reports back auc

class IoTClient(fl.client.NumPyClient):
    def __init__(self, device_name):
        self.device_name = device_name
        self.model = Autoencoder(INPUT_DIM).to(device_torch)
        self.X_train, self.X_test, self.y_test = get_client_data(device_name)
        self.X_train_t = torch.tensor(self.X_train, dtype=torch.float32).to(device_torch)

    def get_parameters(self, config):
        return [v.cpu().numpy() for v in self.model.state_dict().values()]

    def set_parameters(self, parameters):
        params = zip(self.model.state_dict().keys(), parameters)
        state_dict = OrderedDict({k: torch.tensor(v) for k, v in params})
        self.model.load_state_dict(state_dict, strict=True)

    def fit(self, parameters, config):
        self.set_parameters(parameters)
        optimizer = torch.optim.Adam(self.model.parameters(), lr=1e-3)
        criterion = nn.MSELoss()
        self.model.train()
        for _ in range(5):
            optimizer.zero_grad()
            recon = self.model(self.X_train_t)
            loss = criterion(recon, self.X_train_t)
            loss.backward()
            optimizer.step()
        return self.get_parameters({}), len(self.X_train), {}

    def evaluate(self, parameters, config):
        self.set_parameters(parameters)
        self.model.eval()
        with torch.no_grad():
            X_test_t = torch.tensor(self.X_test, dtype=torch.float32).to(device_torch)
            recon = self.model(X_test_t)
            error = torch.mean((X_test_t - recon) ** 2, dim=1).cpu().numpy()
        auc = roc_auc_score(self.y_test, error) if len(set(self.y_test)) > 1 else 0.5
        return 0.0, len(self.X_test), {"roc_auc": auc, "device": self.device_name}

## running the simulation

10 rounds, all 4 devices every round

In [5]:
!pip install -q "flwr[simulation]"

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 73.7/73.7 MB 10.2 MB/s eta 0:00:00


In [6]:
def client_fn(cid):
    return IoTClient(DEVICES[int(cid)]).to_client()

def aggregate_auc(metrics):
    # plain mean over the clients that reported (not example-weighted)
    aucs = [m["roc_auc"] for _, m in metrics]
    devices = [m["device"] for _, m in metrics]
    return {"avg_roc_auc": sum(aucs) / len(aucs),
            "n_reported": len(aucs),
            "per_device_auc": dict(zip(devices, aucs))}

strategy = fl.server.strategy.FedAvg(
    fraction_fit=1.0,
    fraction_evaluate=1.0,
    min_fit_clients=len(DEVICES),
    min_evaluate_clients=len(DEVICES),
    min_available_clients=len(DEVICES),
    evaluate_metrics_aggregation_fn=aggregate_auc,
    fit_metrics_aggregation_fn=lambda ms: {"n_fit_clients": len(ms)},
)

history = fl.simulation.start_simulation(
    client_fn=client_fn,
    num_clients=len(DEVICES),
    config=fl.server.ServerConfig(num_rounds=10),
    strategy=strategy,
    client_resources={"num_cpus": 1},
    ray_init_args={"include_dashboard": False},
)
history

	Instead, use the `flwr run` CLI command to start a local simulation in your Flower app, as shown for example below:

		$ flwr new  # Create a new Flower app from a template

		$ flwr run  # Run the Flower app in Simulation Mode

	Using `start_simulation()` is deprecated.

            This is a deprecated feature. It will be removed
            entirely in future versions of Flower.
        
/usr/local/lib/python3.13/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
	Instead, use the `flwr run` CLI command to start a local simulation in your Flower app, as shown for example below:

		$ flwr new  # Create a new Flower app from a template

		$ flwr run  # Run the Flower app in Simulation Mode

	Using `start_simulation()` is deprecated.

          

History (loss, distributed):
	round 1: 0.0
	round 2: 0.0
	round 3: 0.0
	round 4: 0.0
	round 5: 0.0
	round 6: 0.0
	round 7: 0.0
	round 8: 0.0
	round 9: 0.0
	round 10: 0.0
History (metrics, distributed, fit):
{'n_fit_clients': [(1, 4),
                   (2, 4),
                   (3, 4),
                   (4, 4),
                   (5, 4),
                   (6, 4),
                   (7, 4),
                   (8, 4),
                   (9, 4),
                   (10, 4)]}
History (metrics, distributed, evaluate):
{'avg_roc_auc': [(1, np.float64(0.9999744149013429)),
                 (2, np.float64(0.9999753073318272)),
                 (3, np.float64(0.9999767947467761)),
                 (4, np.float64(0.9999785377042216)),
                 (5, np.float64(0.9999787567484892)),
                 (6, np.float64(0.9999783005617853)),
                 (7, np.float64(0.9999783974903443)),
                 (8, np.float64(0.9999785127938848)),
                 (9, np.float64(0.9999788354234

In [7]:
fit_n = dict(history.metrics_distributed_fit["n_fit_clients"])
eval_n = dict(history.metrics_distributed["n_reported"])
print("clients that trained, per round:     ", [int(fit_n[r]) for r in sorted(fit_n)])
print("clients that reported AUC, per round:", [int(eval_n[r]) for r in sorted(eval_n)])
assert all(v == len(DEVICES) for v in fit_n.values()) and all(v == len(DEVICES) for v in eval_n.values()), \
    "some clients failed: results are not valid for all 4 devices, scroll up for the error"

final = dict(history.metrics_distributed["per_device_auc"])[10]
print({k: round(float(v), 5) for k, v in final.items()},
      "| mean:", round(float(history.metrics_distributed["avg_roc_auc"][-1][1]), 5))

clients that trained, per round:      [4, 4, 4, 4, 4, 4, 4, 4, 4, 4]
clients that reported AUC, per round: [4, 4, 4, 4, 4, 4, 4, 4, 4, 4]
{'Ecobee_Thermostat': 1.0, 'Philips_B120N10_Baby_Monitor': 1.0, 'Danmini_Doorbell': 1.0, 'SimpleHome_XCS7_1002_WHT_Security_Camera': 0.99992} | mean: 0.99998


## federated vs centralized

pulling the centralized numbers back in to compare

In [8]:
import json

with open("/content/drive/MyDrive/nbaiot-project/processed/baseline_results.json") as f:
    baseline = json.load(f)

print("centralized isolation forest:", baseline["isolation_forest"]["roc_auc"])
print("centralized autoencoder:", baseline["autoencoder"]["roc_auc"])
print("federated autoencoder — see history.metrics_distributed above for per-round numbers")

centralized isolation forest: 0.9758805715874938
centralized autoencoder: 0.999988147192871
federated autoencoder — see history.metrics_distributed above for per-round numbers


In [9]:
history.metrics_distributed

{'avg_roc_auc': [(1, np.float64(0.9999744149013429)),
  (2, np.float64(0.9999753073318272)),
  (3, np.float64(0.9999767947467761)),
  (4, np.float64(0.9999785377042216)),
  (5, np.float64(0.9999787567484892)),
  (6, np.float64(0.9999783005617853)),
  (7, np.float64(0.9999783974903443)),
  (8, np.float64(0.9999785127938848)),
  (9, np.float64(0.9999788354234785)),
  (10, np.float64(0.999978876247882))],
 'n_reported': [(1, 4),
  (2, 4),
  (3, 4),
  (4, 4),
  (5, 4),
  (6, 4),
  (7, 4),
  (8, 4),
  (9, 4),
  (10, 4)],
 'per_device_auc': [(1,
   {'Danmini_Doorbell': np.float64(0.9999963357637685),
    'SimpleHome_XCS7_1002_WHT_Security_Camera': np.float64(0.9999050580516475),
    'Philips_B120N10_Baby_Monitor': np.float64(0.9999968933226554),
    'Ecobee_Thermostat': np.float64(0.9999993724673001)}),
  (2,
   {'Philips_B120N10_Baby_Monitor': np.float64(0.9999969176701864),
    'Ecobee_Thermostat': np.float64(0.9999994615091021),
    'Danmini_Doorbell': np.float64(0.9999962123311734),
    

## Results

Final run: Flower 1.39.0, 10 rounds. All 4 clients trained and reported in every round (the check cell above asserts this).

federated reached 0.99998 avg roc-auc across the 4 devices at round 10, against 0.99999 for the
centralized autoencoder in notebook 02 (isolation forest: 0.9759). That gap is far too small to
resolve, so going federated (no raw data leaving each device) didn't cost accuracy on this evaluation.

per device at round 10:
- danmini doorbell: 0.999997
- ecobee thermostat: 1.000000
- philips baby monitor: 0.999998
- simplehome security camera: 0.99992

avg auc was already 0.99997 after round 1 and barely moved (0.999974 to 0.999979), so this
evaluation can't tell how much training is needed. It only shows these 4 attack types are easy.
notebook 04 tests all 10 attack types and finds this was too optimistic for two of them.

camera is consistently the lowest of the four. my guess is its normal traffic is more varied,
but i haven't tested that. notebook 04 also finds the camera among the hardest devices for plain FedAvg.

caveat on comparing with notebook 02: both now use the same benign train/test rows and the same
scaler, but notebook 02 pools all devices into one AUC while this notebook averages per-device AUCs,
and the training budgets differ (30 epochs of minibatches vs 10 rounds x 5 full-batch steps).